_posts/2023-08-29-what-in-the-ml-is-going-on-around-here.md

```diff
@@ -0,0 +1,473 @@
+
+* The engine is one of three. `nmslib`, `faiss` and `lucene`.
+* The `nmslib` engine supports `hnsw`, which is the default if you supply a name but no engine. 
+* The `faiss` engine supports the methods `hnsw` and `ivf`. 
+* The `lucene` engine supports the method `hnsw`.
+
+I'm going to just specify a method name of `hnsw` which will provide the default engine of `nmslib`. Here's the call to create my KNN index now.
+```json
+PUT /super-awesome-nlp-index
+{
+  "settings": {
+    "index.knn": true,
+    "default_pipeline": "i-eat-pieces-of-nlp-pipelines-for-breakfast"
+  },
+  "mappings": {
+    "properties": {
+      "passage_embedding": {
+        "type": "knn_vector",
+        "dimension": 384,
+        "method": {
+          "name": "hnsw"
+        }
+      },
+      "passage_text": {
+        "type": "text"
+      }
+    }
+  }
+}
+
+# The Response
+{
+    "acknowledged": true,
+    "shards_acknowledged": true,
+    "index": "super-awesome-nlp-index"
+}
+```
+
+Double rad! I made a KNN index attached to my neural search pipeline! NOW where was I? Oh right! I just want to ingest and vectorize some text, and that's it. I feel like I'm getting close. 
+
+```json
+POST /super-awesome-nlp-index/_doc
+{
+   "passage_text": "Why did the chicken cross the road? To get to the other side."
+}
+
+# The Response
+{
+    "_index": "super-awesome-nlp-index",
+    "_id": "SOPsmYoBKue4OlrZShCY",
+    "_version": 1,
+    "result": "created",
+    "_shards": {
+        "total": 2,
+        "successful": 1,
+        "failed": 0
+    },
+    "_seq_no": 0,
+    "_primary_term": 1
+}
+```
+
+Looks like it ingested my joke. Just to see it with my own eyes I'm going to do a quick search and see if the vector comes back. I'll point out again that the "dimensionality" of our vector is 384, so there's going to be 384 values. I'm going to cut some of them out. 
+
+```json
+{
+  "took": 0,
+  "timed_out": false,
+  "_shards": {
+    "total": 1,
+    "successful": 1,
+    "skipped": 0,
+    "failed": 0
+  },
+  "hits": {
+    "total": {
+      "value": 1,
+      "relation": "eq"
+    },
+    "max_score": 1,
+    "hits": [
+      {
+        "_index": "super-awesome-nlp-index",
+        "_id": "SOPsmYoBKue4OlrZShCY",
+        "_score": 1,
+        "_source": {
+          "passage_embedding": [
+            0.09026472,
+            0.022673616,
+            -0.027812438,
+            0.09684542,
+            0.083259515,
+            0.04510128,
+            0.05745254,
+            # SNIP!
+          ],
+          "passage_text": "Why did the chicken cross the road? To get to the other side."
+        }
+      }
+    ]
+  }
+}
+```
+This is where I stop. The goal has been achieved - a KNN index associated with a neural search pipeline and an ML model. Please stay tuned - I have no big plans for now, but where dad jokes are concerned, the sky is the limit. 
```

Comment:
I'm actually not sure whether the concept of "dad jokes" translates across cultures. I would probably just use "bad" instead to avoid any cultural mistranslation (jokes literally about dads). Also, I would add a little more of a conclusion here. What do you hope the reader learned as a result of reading this post, and where else can they go to learn more or to provide feedback? In technical writing, we say that you should "tell them what you're going to tell them, then tell them, then tell them what you told them."